### ISSUE TITLE
[PATCH] Coverity: va_args/va_copy without va_end

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13778]
Migrated from [rt.perl.org#121748](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=121748) (status was 'resolved')

Searchable as RT121748$

**Comment by p5pRT** [comment]
### From @jhi

va\_end is not optional\.

Attached\.

**Comment by p5pRT** [comment]
### From @jhi

<details><summary>0001-Fix-for-Coverity-perl5-CIDs-29225-29226-29227-29228-.patch</summary>

```diff
From 74a2420d94fc4ac200f7cbd4b6f9963ec7d2ccb6 Mon Sep 17 00:00:00 2001
From: Jarkko Hietaniemi <jhi@iki.fi>
Date: Wed, 23 Apr 2014 12:53:42 -0400
Subject: [PATCH] Fix for Coverity perl5 CIDs 29225, 29226, 29227, 29228,
 29229: Missing varargs init or cleanup (VARARGS) missing va_end: va_end was
 not called for foo.

Use of va_args must be finished off with va_end (in other words,
use of va_start or va_copy must be bracketed off with va_end).
In most platforms va_end is a no-op, but in some platforms it is
required for proper cleanup (or face stack smash, or memory leak).
---
 ext/PerlIO-via/via.xs |  8 ++++----
 mathoms.c             | 10 ++++++++--
 perlio.c              |  1 +
 util.c                |  1 +
 4 files changed, 14 insertions(+), 6 deletions(-)

diff --git a/ext/PerlIO-via/via.xs b/ext/PerlIO-via/via.xs
index 619174a..d0f3d78 100644
--- a/ext/PerlIO-via/via.xs
+++ b/ext/PerlIO-via/via.xs
@@ -68,12 +68,12 @@ PerlIOVia_method(pTHX_ PerlIO * f, const char *method, CV ** save, int flags,
 		 ...)
 {
     PerlIOVia *s = PerlIOSelf(f, PerlIOVia);
+    SV *result = Nullsv;
     CV *cv =
 	(*save) ? *save : PerlIOVia_fetchmethod(aTHX_ s, method, save);
-    SV *result = Nullsv;
-    va_list ap;
-    va_start(ap, flags);
     if (cv != (CV *) - 1) {
+        va_list ap;
+        va_start(ap, flags);
 	IV count;
 	dSP;
 	SV *arg;
@@ -84,6 +84,7 @@ PerlIOVia_method(pTHX_ PerlIO * f, const char *method, CV ** save, int flags,
 	while ((arg = va_arg(ap, SV *))) {
 	    XPUSHs(arg);
 	}
+        va_end(ap);
 	if (*PerlIONext(f)) {
 	    if (!s->fh) {
 		GV *gv;
@@ -121,7 +122,6 @@ PerlIOVia_method(pTHX_ PerlIO * f, const char *method, CV ** save, int flags,
 	LEAVE;
 	POPSTACK;
     }
-    va_end(ap);
     return result;
 }
 
diff --git a/mathoms.c b/mathoms.c
index 73f1e8d..1132ac8 100644
--- a/mathoms.c
+++ b/mathoms.c
@@ -544,6 +544,7 @@ int
 Perl_fprintf_nocontext(PerlIO *stream, const char *format, ...)
 {
     dTHXs;
+    int ret = 0;
     va_list(arglist);
 
     /* Easier to special case this here than in embed.pl. (Look at what it
@@ -553,7 +554,9 @@ Perl_fprintf_nocontext(PerlIO *stream, const char *format, ...)
 #endif
 
     va_start(arglist, format);
-    return PerlIO_vprintf(stream, format, arglist);
+    ret = PerlIO_vprintf(stream, format, arglist);
+    va_end(arglist);
+    return ret;
 }
 
 int
@@ -561,13 +564,16 @@ Perl_printf_nocontext(const char *format, ...)
 {
     dTHX;
     va_list(arglist);
+    int ret = 0;
 
 #ifdef PERL_IMPLICIT_CONTEXT
     PERL_ARGS_ASSERT_PRINTF_NOCONTEXT;
 #endif
 
     va_start(arglist, format);
-    return PerlIO_vprintf(PerlIO_stdout(), format, arglist);
+    ret = PerlIO_vprintf(PerlIO_stdout(), format, arglist);
+    va_end(arglist);
+    return ret;
 }
 
 #if defined(HUGE_VAL) || (defined(USE_LONG_DOUBLE) && defined(HUGE_VALL))
diff --git a/perlio.c b/perlio.c
index 0ae0a43..8d8c873 100644
--- a/perlio.c
+++ b/perlio.c
@@ -4915,6 +4915,7 @@ PerlIO_vprintf(PerlIO *f, const char *fmt, va_list ap)
     s = SvPV_const(sv, len);
     wrote = PerlIO_write(f, s, len);
     SvREFCNT_dec(sv);
+    va_end(apc);
     return wrote;
 }
 
diff --git a/util.c b/util.c
index 0a0ee40..09b2679 100644
--- a/util.c
+++ b/util.c
@@ -4935,6 +4935,7 @@ Perl_my_vsnprintf(char *buffer, const Size_t len, const char *format, va_list ap
     retval = vsprintf(buffer, format, ap);
 # endif
 #endif /* #ifdef NEED_VA_COPY */
+    va_end(apc);
     /* vsprintf() shows failure with < 0 */
     if (retval < 0
 #ifdef HAS_VSNPRINTF
-- 
1.9.2


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Sat Apr 26 13&#8203;:04&#8203;:36 2014\, jhi wrote&#8203;:
> va\_end is not optional\.
> 
> Attached\.


It turns out the va\_end\(\) in this chunk&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/util.c b/util.c
index 0a0ee40..09b2679 100644
--- a/util.c
+++ b/util.c
@@ -4935,6 +4935,7 @@ Perl_my_vsnprintf(char *buffer, const Size_t len, const char *format, va_list ap
     retval = vsprintf(buffer, format, ap);
 # endif
 #endif /* #ifdef NEED_VA_COPY */
+    va_end(apc);
     /* vsprintf() shows failure with < 0 */
     if (retval < 0
 #ifdef HAS_VSNPRINTF


```
</details>
is in the wrong place, producing a build error on Win32.

Tony

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jhi

On Wednesday\-201405\-21\, 20&#8203;:35\, Tony Cook via RT wrote&#8203;:
> is in the wrong place\, producing a build error on Win32\.

Smoke link?

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed May 21 17&#8203;:37&#8203;:51 2014\, jhi wrote&#8203;:
> On Wednesday\-201405\-21\, 20&#8203;:35\, Tony Cook via RT wrote&#8203;:
> > is in the wrong place\, producing a build error on Win32\.
> 
> Smoke link?
> 


http&#8203;://www\.nntp\.perl\.org/group/perl\.daily\-build\.reports/2014/05/msg161602\.html

Same problem in perlio\.c\.

apc is only declared inside \#ifdef NEED\_VA\_COPY\, but the va\_end\(\) it outside of it\.

I'll push some temp fixes for this and other issues soon\.

Tony

**Comment by p5pRT** [comment]
### From @jhi

On Wednesday\-201405\-21\, 21&#8203;:54\, Tony Cook via RT wrote&#8203;:
> apc is only declared inside \#ifdef NEED\_VA\_COPY\, but the va\_end\(\) it outside of it\.

Ahh\, duh\.

> I'll push some temp fixes for this and other issues soon\.

Thanks\!

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed May 21 19&#8203;:03&#8203;:20 2014\, jhi wrote&#8203;:
> On Wednesday\-201405\-21\, 21&#8203;:54\, Tony Cook via RT wrote&#8203;:
> > apc is only declared inside \#ifdef NEED\_VA\_COPY\, but the va\_end\(\) it
> > outside of it\.
> 
> Ahh\, duh\.
> 
> > I'll push some temp fixes for this and other issues soon\.
> 
> Thanks\!

Another&#8203;:

\-\-\- a/ext/PerlIO\-via/via\.xs
\+\+\+ b/ext/PerlIO\-via/via\.xs
@&#8203;@&#8203; \-68\,12 \+68\,12 @&#8203;@&#8203; PerlIOVia\_method\(pTHX\_ PerlIO \* f\, const char \*method\, CV \*\* save\, int flags\,
&nbsp;          \.\.\.\)
 \{
&nbsp;     PerlIOVia \*s = PerlIOSelf\(f\, PerlIOVia\);
\+    SV \*result = Nullsv;
&nbsp;     CV \*cv =
&nbsp;     \(\*save\) ? \*save : PerlIOVia\_fetchmethod\(aTHX\_ s\, method\, save\);
\-    SV \*result = Nullsv;
\-    va\_list ap;
\-    va\_start\(ap\, flags\);
&nbsp;     if \(cv \!= \(CV \*\) \- 1\) \{
\+        va\_list ap;
\+        va\_start\(ap\, flags\);
&nbsp;     IV count;
&nbsp;     dSP;
&nbsp;     SV \*arg;

has code \(va\_start\(\)\) before declarations\, breaking C89 builds\.

Tony

**Comment by p5pRT** [comment]
### From @jhi

On Wednesday\-201405\-21\, 22&#8203;:07\, Tony Cook via RT wrote&#8203;:
> has code \(va\_start\(\)\) before declarations\, breaking C89 builds\.

Grrr\.  I think I will start adding \-std=c89 to my builds\.\.\.

In this particular case the va\_start\(\) can be pushed down after all the 
decls\, no?

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed\, May 21\, 2014 at 10&#8203;:15&#8203;:36PM \-0400\, Jarkko Hietaniemi wrote&#8203;:
> On Wednesday\-201405\-21\, 22&#8203;:07\, Tony Cook via RT wrote&#8203;:
> >has code \(va\_start\(\)\) before declarations\, breaking C89 builds\.
> 
> Grrr\.  I think I will start adding \-std=c89 to my builds\.\.\.
> 
> In this particular case the va\_start\(\) can be pushed down after all
> the decls\, no?

Right\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

On Sat Apr 26 13&#8203;:04&#8203;:36 2014\, jhi wrote&#8203;:
> va\_end is not optional\.
> 
> Attached\.

Thanks\, applied as 3ed3a8afebd64616aef147205403b96b30a4b4ee with the declaration issue fixed\, but I forget the cpp nesting fix\, which is in d4825b278e28006bdc9c3f36ab174eade62d6c4c\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'
